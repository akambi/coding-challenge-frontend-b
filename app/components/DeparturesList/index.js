import React from 'react';
import PropTypes from 'prop-types';

import List from 'components/List';
import ListItem from 'components/ListItem';
import LoadingIndicator from 'components/LoadingIndicator';
import DepartureListItem from 'containers/DepartureListItem';

function DeparturesList({ loading, error, departures }) {
  if (loading) {
    return <List component={LoadingIndicator} />;
  }

  if (error !== false) {
    const ErrorComponent = () => (
      <ListItem item="Something went wrong, please try again!" />
    );
    return <List component={ErrorComponent} />;
  }

  if (departures !== false) {
    return <List items={departures} component={DepartureListItem} />;
  }

  return null;
}

DeparturesList.propTypes = {
  loading: PropTypes.bool,
  error: PropTypes.any,
  departures: PropTypes.any,
};

export default DeparturesList;
