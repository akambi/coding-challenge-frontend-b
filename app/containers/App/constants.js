export const QUEBEC = 'quebec';
export const MONTREAL = 'montreal';
export const API_KEY = 'PARTNER_BaASYYHxTxuOINEOMWq5GA';

export const CityToGeoHashMapping = {
  quebec: 'f2m673',
  montreal: 'f25dvk',
};

export const CHANGE_SEARCH_PARAMS = 'busbud/App/CHANGE_SEARCH_PARAMS';

export const LOAD_DEPARTURES = 'busbud/App/LOAD_DEPARTURES';
export const LOAD_DEPARTURES_SUCCESS = 'busbud/App/LOAD_DEPARTURES_SUCCESS';
export const LOAD_DEPARTURES_ERROR = 'busbud/App/LOAD_DEPARTURES_ERROR';
