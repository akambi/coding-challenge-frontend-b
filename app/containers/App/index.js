import React from 'react';
import { Helmet } from 'react-helmet';
import styled from 'styled-components';
import { Switch, Route } from 'react-router-dom';

import HomePage from 'containers/HomePage/Loadable';
import NotFoundPage from 'containers/NotFoundPage/Loadable';
import Header from 'components/Header';
import Footer from 'components/Footer';

import GlobalStyle from '../../global-styles';

const AppWrapper = styled.div`
  max-width: calc(768px + 16px * 2);
  margin: 0 auto;
  display: flex;
  min-height: 100%;
  padding: 0 16px;
  flex-direction: column;
`;

export default function App() {
  return (
    <AppWrapper>
      <Helmet
        titleTemplate="%s - Busbud's bus departures"
        defaultTitle="Busbud's bus departures"
      >
        <meta
          name="description"
          content="A simple application that lists all departures for a given origin city and a given destination city for a given day"
        />
      </Helmet>
      <Header />
      <Switch>
        <Route exact path="/" component={HomePage} />
        <Route path="" component={NotFoundPage} />
      </Switch>
      <Footer />
      <GlobalStyle />
    </AppWrapper>
  );
}
