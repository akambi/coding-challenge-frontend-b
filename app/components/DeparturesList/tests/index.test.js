import React from 'react';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import { browserHistory } from 'react-router-dom';
import { render } from 'react-testing-library';

import DeparturesList from '../index';
import configureStore from '../../../configureStore';
import { MONTREAL, QUEBEC } from '../../../containers/App/constants';

describe('<DeparturesList />', () => {
  it('should render the loading indicator when its loading', () => {
    const { container } = render(<DeparturesList loading />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render an error if loading failed', () => {
    const { queryByText } = render(
      <IntlProvider locale="en">
        <DeparturesList
          loading={false}
          error={{ message: 'Loading failed!' }}
        />
      </IntlProvider>,
    );
    expect(queryByText(/Something went wrong/)).not.toBeNull();
  });

  it('should render the departures if loading was successful', () => {
    const store = configureStore(
      {
        global: {
          searchParams: {
            origin: QUEBEC,
            destination: MONTREAL,
            outboundDate: '2020-12-14T14:21:30.490Z',
            adult: 1,
          },
        },
      },
      browserHistory,
    );
    const departures = [
      {
        departure_time: '2020-12-13 09:10:00',
        arrival_time: '2020-12-13 14:00:00',
        location: {
          id: 1,
          name: 'quebec',
        },
        price: 5520,
      },
    ];
    const { container } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <DeparturesList departures={departures} error={false} />
        </IntlProvider>
      </Provider>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should not render anything if nothing interesting is provided', () => {
    const { container } = render(
      <DeparturesList departures={false} error={false} loading={false} />,
    );

    expect(container.firstChild).toBeNull();
  });
});
