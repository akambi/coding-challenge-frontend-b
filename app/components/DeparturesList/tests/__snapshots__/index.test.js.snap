// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DeparturesList /> should render the departures if loading was successful 1`] = `
<div
  class="Wrapper-sc-1l7ow4n-0 gikgZI"
>
  <ul
    class="Ul-sc-1l9y7is-0 jTPOp"
  >
    <li
      class="Wrapper-sc-11n0iqq-0 fJThNy"
    >
      <div
        class="Item-sc-141m8an-0 gvGHMp"
      >
        <div
          class="Wrapper-sc-1k2q7aa-0 dtddjJ"
        >
          <div>
            <strong>
              09:10:00
               
            </strong>
            <span>
              Québec
            </span>
            <span>
               - 
              quebec
            </span>
          </div>
          <div>
            <strong>
              14:00:00
               
            </strong>
            <span>
              Montréal
            </span>
          </div>
          <div>
            <strong>
              <span>
                CA$55.20
              </span>
            </strong>
          </div>
        </div>
      </div>
    </li>
  </ul>
</div>
`;

exports[`<DeparturesList /> should render the loading indicator when its loading 1`] = `
<div
  class="Wrapper-sc-1l7ow4n-0 gikgZI"
>
  <ul
    class="Ul-sc-1l9y7is-0 jTPOp"
  >
    <div
      class="Wrapper-lo97gk-0 eKejYq"
    >
      <div
        class="Circle__CirclePrimitive-rqgep1-0 dVhQYz"
      />
      <div
        class="Circle__CirclePrimitive-rqgep1-0 tXKYg"
      />
      <div
        class="Circle__CirclePrimitive-rqgep1-0 dsPEef"
      />
      <div
        class="Circle__CirclePrimitive-rqgep1-0 kupaCF"
      />
      <div
        class="Circle__CirclePrimitive-rqgep1-0 ctNzoM"
      />
      <div
        class="Circle__CirclePrimitive-rqgep1-0 kkwFVw"
      />
      <div
        class="Circle__CirclePrimitive-rqgep1-0 jByibz"
      />
      <div
        class="Circle__CirclePrimitive-rqgep1-0 kMisqh"
      />
      <div
        class="Circle__CirclePrimitive-rqgep1-0 gjakJJ"
      />
      <div
        class="Circle__CirclePrimitive-rqgep1-0 JfqmL"
      />
      <div
        class="Circle__CirclePrimitive-rqgep1-0 kJVFAj"
      />
      <div
        class="Circle__CirclePrimitive-rqgep1-0 fIjItU"
      />
    </div>
  </ul>
</div>
`;
